homeassistant/components/upcloud.py

```diff
@@ -0,0 +1,170 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_CORE_NUMBER = 'core_number'
+ATTR_HOSTNAME = 'hostname'
+ATTR_MEMORY_AMOUNT = 'memory_amount'
+ATTR_STATE = 'state'
+ATTR_TITLE = 'title'
+ATTR_UUID = 'uuid'
+ATTR_ZONE = 'zone'
+
+CONF_SERVERS = 'servers'
+
+DATA_UPCLOUD = 'data_upcloud'
+DOMAIN = 'upcloud'
+
+DEFAULT_COMPONENT_NAME = 'UpCloud {}'
+DEFAULT_COMPONENT_DEVICE_CLASS = 'power'
+
+UPCLOUD_PLATFORMS = ['binary_sensor', 'switch']
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+SIGNAL_UPDATE_UPCLOUD = "upcloud_update"
+
+STATE_MAP = {
+    "started": STATE_ON,
+    "stopped": STATE_OFF,
+    "error": STATE_PROBLEM,
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL):
+            cv.time_period,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the UpCloud component."""
+    import upcloud_api
+
+    conf = config[DOMAIN]
+    username = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASSWORD)
+    scan_interval = conf.get(CONF_SCAN_INTERVAL)
+
+    manager = upcloud_api.CloudManager(username, password)
+
+    try:
+        manager.authenticate()
+        hass.data[DATA_UPCLOUD] = UpCloud(manager)
+    except upcloud_api.UpCloudAPIError:
+        _LOGGER.error("Authentication failed.")
+        return False
+
+    def upcloud_update(event_time):
+        """Call UpCloud to update information."""
+        _LOGGER.debug("Updating UpCloud component")
+        hass.data[DATA_UPCLOUD].update()
+        dispatcher_send(hass, SIGNAL_UPDATE_UPCLOUD)
+
+    # Call the UpCloud API to refresh data
+    track_time_interval(hass, upcloud_update, scan_interval)
+
+    return True
+
+
+class UpCloud(object):
+    """Handle all communication with the UpCloud API."""
+
+    def __init__(self, manager):
+        """Initialize the UpCloud connection."""
+        self.data = {}
+        self.manager = manager
+
+    def update(self):
+        """Update data from UpCloud API."""
+        self.data = {
+            server.uuid: server for server in self.manager.get_servers()
+        }
+
+
+class UpCloudServerEntity(Entity):
+    """Entity class for UpCloud servers."""
+
+    def __init__(self, upcloud, uuid):
+        """Initialize the UpCloud server entity."""
+        self._upcloud = upcloud
+        self.uuid = uuid
+        self.data = None
+
+    @property
+    def name(self):
+        """Return the name of the component."""
+        try:
+            return DEFAULT_COMPONENT_NAME.format(self.data.title)
+        except (AttributeError, KeyError, TypeError):
+            return DEFAULT_COMPONENT_NAME.format(self.uuid)
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        async_dispatcher_connect(
+            self.hass, SIGNAL_UPDATE_UPCLOUD, self._update_callback)
+
+    def _update_callback(self):
```

Comment:
This could be made async by annotating the method with `@callback` from core.py.